provider/core/src/refcount.rs

```diff
@@ -0,0 +1,222 @@
+use core::ptr::NonNull;
+use yoke::CloneableCart;
+
+#[cfg(not(feature = "sync"))]
+pub(crate) use alloc::rc::Rc as SelectedRc;
+#[cfg(feature = "sync")]
+pub(crate) use alloc::sync::Arc as SelectedRc;
+
+type WrappedType = Box<[u8]>;
+
+const SENTINEL_PTR: *mut WrappedType = core::mem::align_of::<WrappedType>() as *mut WrappedType;
+
+/// A thin wrapper over [`NonNull`] that works with `*const T`
+#[derive(Debug)]
+#[repr(transparent)]
+pub(crate) struct OptionSelectedRcBytes(NonNull<WrappedType>);
+
+impl OptionSelectedRcBytes {
+    #[inline]
+    pub const fn new_sentinel() -> Self {
+        // Safety: `SENTINEL_PTR` is non-null
+        Self(unsafe { NonNull::new_unchecked(SENTINEL_PTR) })
+    }
+
+    #[inline]
+    pub fn from_selected_rc(selected_rc: SelectedRc<WrappedType>) -> Self {
+        let ptr = SelectedRc::into_raw(selected_rc);
+        debug_assert_ne!(
+            ptr, SENTINEL_PTR,
+            "Creating from [A]Rc should never be the sentinel ptr"
+        );
+        debug_assert_ne!(
+            ptr, 0 as *const WrappedType,
+            "Creating from [A]Rc should never be the null ptr"
+        );
+        // Safety: ptr is non-null
+        Self(unsafe { NonNull::new_unchecked(ptr as *mut WrappedType) })
+    }
+
+    #[inline]
+    pub fn take(&mut self) -> Option<SelectedRc<WrappedType>> {
+        let ptr = self.0.as_ptr() as *const WrappedType;
+        // Safety: `SENTINEL_PTR` is non-null
+        self.0 = unsafe { NonNull::new_unchecked(SENTINEL_PTR) };
+        if ptr == SENTINEL_PTR {
+            None
+        } else {
+            // Safety: this type has only 2 constructors, and we already
+            // checked for the sentinel value.
+            Some(unsafe { SelectedRc::from_raw(ptr) })
+        }
+    }
+
+    pub fn as_bytes(&self) -> Option<&[u8]> {
+        let ptr = self.0.as_ptr() as *const WrappedType;
+        if ptr == SENTINEL_PTR {
+            None
+        } else {
+            // Safety: this type has only 2 constructors, and we already
+            // checked for the sentinel value.
+            let wrapped_type: &WrappedType = unsafe { self.0.as_ref() };
+            Some(&*wrapped_type)
+        }
+    }
+}
+
+impl Drop for OptionSelectedRcBytes {
+    fn drop(&mut self) {
+        self.take();
+    }
+}
+
+impl Clone for OptionSelectedRcBytes {
+    fn clone(&self) -> Self {
+        let ptr = self.0.as_ptr() as *const WrappedType;
+        if ptr == SENTINEL_PTR {
+            Self::new_sentinel()
+        } else {
+            // Safety: this type has only 2 constructors, and we already
+            // checked for the sentinel value.
+            let selected_rc = unsafe { SelectedRc::from_raw(ptr) };
+            // Increase the ref count
+            let _ = ManuallyDrop::new(selected_rc.clone());
+            Self::from_selected_rc(selected_rc)
+        }
+    }
+}
+
+// Safety: type has the same semantics as Option<Rc<_>>
+// which implements CloneableCart
```

Comment:
nit: I'd have a duplicate of the safety reasoning here as well

Reply:
CloneableCart says:

> Safety
>
> This trait is safe to implement StableDeref types which, once Cloned, point to the same underlying data.
>
> (This trait is also implemented on Option<T> and (), which are the two non-StableDeref cart types that Yokes can be constructed for)

But this type does not implement Deref and therefore cannot implement StableDeref. It is an optional deref; it can yield `Option<&Deref>` but not `&Deref`. Should I change the safety requirements of `CloneableCart` itself to allow for this?

Reply:
Yeah we should update the docs to describe it as okay to implement on certain other kinds of types. Might take some work, mostly saying that it is for reference counted types where any data that is referenceable as a cart is still owned by the new clones reference.